apps/src/lib/kits/maker/boards/VirtualMBBoard.js

```diff
@@ -0,0 +1,225 @@
+/** @file For running Maker apps without an attached MicroBit board. */
+import {EventEmitter} from 'events'; // provided by webpack's node-libs-browser
+import {MBFirmataClientStub} from '../util/makeStubBoard';
+
+/**
+ * Virtual MicroBit Board for running Maker Toolkit apps without a MicroBit board
+ * attached.
+ * Attaches virtual, no-op components to the interpreter.
+ * @extends EventEmitter
+ * @implements MakerBoard
+ */
+export default class VirtualMBBoard extends EventEmitter {
+  constructor() {
+    super();
+    this.boardClient_ = new MBFirmataClientStub();
+  }
+
+  /**
+   * Open a connection to the board on its configured port.
+   * @returns {Promise} resolved when the board is ready to use.
+   */
+  connect() {
+    return Promise.resolve();
+  }
+
+  /**
+   * Disconnect and clean up the board controller and all components.
+   * @returns {Promise}
+   */
+  destroy() {
+    return Promise.resolve();
+  }
+
+  reset() {}
+
+  /**
+   * Create a serial port controller and open the serial port immediately.
+   * @return {SerialPort}
+   */
+  openSerialPort() {}
+
+  /**
+   * Marshals the board component controllers and appropriate constants into the
+   * given JS Interpreter instance so they can be used by student code.
+   *
+   * @param {JSInterpreter} jsInterpreter
+   */
+  installOnInterpreter(jsInterpreter) {
+    const constructors = {
+      MicroBitButton: VirtualMicroBitButton,
+      LedScreen: VirtualLedScreen,
+      Accelerometer: VirtualAccelerometer,
+      MicroBitThermometer: VirtualMicroBitThermometer,
+      Compass: VirtualCompass,
+      LightSensor: VirtualLightSensor,
+      ExternalButton: VirtualExternalButton,
+      ExternalLed: VirtualExternalLed,
+      CapacitiveTouchSensor: VirtualCapacitiveTouchSensor
+    };
+
+    for (const constructorName in constructors) {
+      const constructor = constructors[constructorName];
+      jsInterpreter.addCustomMarshalObject({instance: constructor});
+      jsInterpreter.createGlobalProperty(constructorName, constructor);
+    }
+
+    const components = {
+      buttonA: new VirtualMicroBitButton(),
+      buttonB: new VirtualMicroBitButton(),
+      ledScreen: new VirtualLedScreen(),
+      tempSensor: new VirtualMicroBitThermometer(),
+      accelerometer: new VirtualAccelerometer(),
+      lightSensor: new VirtualLightSensor(),
+      compass: new VirtualCompass(),
+      board: new VirtualMBFirmataWrapper()
+    };
+
+    for (const componentName in components) {
+      const component = components[componentName];
+      jsInterpreter.createGlobalProperty(componentName, component);
+    }
+  }
+
+  /**
+   * @param {number} pin
+   * @param {number} modeConstant
+   */
+  pinMode(pin, modeConstant) {}
```

Comment:
Ooo - I like that idea. I also want to note that we talked about adding user console statements for the virtual board's function calls so the user can see that the calls were made successfully since we don't have a board simulator (yet).
I'll create a ticket to do both of these things.

Reply:
https://codedotorg.atlassian.net/browse/SL-773